client/src/api/v2/camera.ts

```diff
@@ -0,0 +1,294 @@
+  recordingMinutes: Number,
+  /** Path to video file being recorded */
+  recordingFile: String,
+  diskSpaceRemaining: Number,
+});
+const CameraRecordingErrorPayload = Record({
+  status: Literal('error'),
+  diskSpaceRemaining: Number,
+  /** Description of error that has occurred */
+  error: String,
+});
+const CameraRecordingStatusPayload = Union(
+  CameraRecordingOffPayload,
+  CameraRecordingOnPayload,
+  CameraRecordingErrorPayload,
+);
+type CameraRecordingStatusPayload = Static<typeof CameraRecordingStatusPayload>;
+
+interface CameraRecordingStatusItem {
+  /** Display name of the status item e.g. "Disk Space Remaining" */
+  name: string;
+  /** Value of the status item e.g. "1.2 GiB" */
+  value: string;
+}
+
+/**
+ * Parse the recording payload into an object and format the values
+ *
+ * Payload structure is defined in the 'V3 MQTT Topics' page on Notion.
+ * Topic is /v3/camera/recording/status/<primary/>secondary>.
+ *
+ * @param payload Payload from MQTT message, parsed
+ * @returns Formatted payload
+ */
+export function formatRecordingPayload(payload?: CameraRecordingStatusPayload) {
+  if (!payload) return null;
+
+  // Format data always present regardless of status
+  const commonData: CameraRecordingStatusItem[] = [
+    { name: 'Status', value: capitalise(payload.status) },
+    {
+      name: 'Disk space remaining',
+      value: formatBytes(payload.diskSpaceRemaining),
+    },
+  ];
+
+  // Format data that is dependent on status
+  const formatter = CameraRecordingStatusPayload.match(
+    () => [],
+    (onPayload) => [
+      {
+        name: 'Recording duration',
+        value: formatMinutes(onPayload.recordingMinutes),
+      },
+      { name: 'Recording file', value: onPayload.recordingFile },
+    ],
+    (errorPayload) => [{ name: 'Error', value: errorPayload.error }],
+  );
+
+  return commonData.concat(formatter(payload));
+}
+
+/**
+ * Initiate receiving status payloads
+ *
+ * @param subComponent The sub component to init for (e.g. recording, video feed)
+ * @param device Device
+ */
+function initStatus(subComponent: string, device?: CameraDevice) {
+  const path = ['camera', subComponent];
+  if (device) path.push(device);
+  emit(`get-status-payload`, path);
+}
+
+/**
+ * Makes the device name pretty.
+ *
+ * Hardcoded for efficiency. If you don't like it, complain to Angus Trau :).
+ *
+ * @param device Device
+ * @returns  Prettied device name
+ */
+export function getPrettyDeviceName(device: CameraDevice) {
+  return device === 'primary' ? 'Primary' : 'Secondary';
+}
+
+interface StatusPayloadOptions<T> {
+  /** Initial value for the payload */
+  initValue: T | null;
+  /** Handler for updating payload */
+  payloadHandler?: (
+    setter: React.Dispatch<React.SetStateAction<T | null>>,
+    newPayload: T | null,
+    device: CameraDevice,
+  ) => T | void;
+  /** Handler for return value */
+  returnHandler?: (payload: T | null, device?: CameraDevice) => T | null;
```

Comment:
There are actually two types here:

- `T` the type of the socket.io payload
- `U` the type of the value returned by `returnHandler` since `returnHandler` can handle the stored state value however it wants.